google-ads-examples/src/main/java/com/google/ads/googleads/examples/accountmanagement/GetAccountHierarchy.java

```diff
@@ -0,0 +1,296 @@
+          long customer = Long.parseLong(CustomerName.parse(customerResourceName).getCustomer());
+          System.out.println(customer);
+          seedCustomerIds.add(customer);
+        }
+      }
+    }
+
+    for (long seedCustomerId : seedCustomerIds) {
+      Set<Long> managerAccountsToSearch = new HashSet<>();
+      CustomerClient rootCustomerClient = null;
+      int depth = 0;
+
+      if (loginCustomerId != null) {
+        googleAdsClient = createGoogleAdsClient(loginCustomerId);
+      } else {
+        googleAdsClient = createGoogleAdsClient(seedCustomerId);
+      }
+
+      if (googleAdsClient == null) {
+        return;
+      }
+
+      // Creates the Google Ads Service client.
+      try (GoogleAdsServiceClient googleAdsServiceClient =
+          googleAdsClient.getLatestVersion().createGoogleAdsServiceClient()) {
+        // Creates a query that retrieves all child accounts of the manager specified in search
+        // calls below.
+        String query =
+            "SELECT customer_client.client_customer, customer_client.level, "
+                + "customer_client.manager, customer_client.descriptive_name, "
+                + "customer_client.currency_code, customer_client.time_zone, "
+                + "customer_client.id "
+                + "FROM customer_client "
+                + "WHERE customer_client.level <= 1";
+
+        // Adds the seed customer ID to the list of IDs to be processed.
+        managerAccountsToSearch.add(seedCustomerId);
+        // Performs a breadth-first search algorithm to build a mapping of managers to their
+        // child accounts.
+        Multimap<Long, CustomerClient> customerIdsToChildAccounts = ArrayListMultimap.create();
+        while (!managerAccountsToSearch.isEmpty()) {
+          long customerId = managerAccountsToSearch.iterator().next();
+          managerAccountsToSearch.remove(customerId);
+          SearchPagedResponse response = null;
+          try {
+            // Issues a search request.
+            response =
+                googleAdsServiceClient.search(
+                    SearchGoogleAdsRequest.newBuilder()
+                        .setQuery(query)
+                        .setCustomerId(Long.toString(customerId))
+                        .build());
+
+            // Iterates over all rows in all pages to get all customer clients under the specified
+            // customer's hierarchy.
+            for (GoogleAdsRow googleAdsRow : response.iterateAll()) {
+              CustomerClient customerClient = googleAdsRow.getCustomerClient();
+              // The customer client making the request will be returned in the
+              // search result with a level of 0.
+              if (customerClient.getLevel().getValue() == 0) {
+                // Store the root customer client, which is the first encountered customer client
+                // that has level of 0.
+                if (rootCustomerClient == null) {
```

Comment:
Can you explain why we need to get this from the API? Can't we just store the first customer ID we requested the customer clients for?

Reply:
We technically don't need this to find the root customer ID. This will simply be the seedCustomerId. We do, however, need two checks. We need to check if the current customer is the local root, in which case we'll continue - that's the if statement outside of this one. We also need to check if the current customer is the global root (or seed ID) so that we can save the customerClient to a local variable because we need the full object (not just the ID) for the printAccountHierarchy method. This would be the first time we'd have access to the full object. If it's provided as a parameter, it's only provided as an ID. If retrieved from listAccessibleCustomers, we'll still only have the resources names.

We can refactor the nested if statements as follows to make it clearer.

```
              if (customerClient.getId().getValue() == seedCustomerId) {
                rootCustomerClient = customerClient;
              }

              // The steps below map parent and children accounts. Continue here so that managers
              // accounts exclude themselves from the list of their children accounts.
              if (customerClient.getId().getValue() == customerId) {
                continue;
              }
```

@fiboknacky you wrote this in the PHP golden, so please let me know if I've missed anything or if you have any input.

Reply:
Also @nwbirnie - not sure if this was your question, but the results from the search query are not guaranteed to come back in order, so the local root may not be the first result.